### PR TITLE
Policy attributes design proposal

### PR DESCRIPTION
**Pull request by elevran**
See also Issue #17

### CODE REVIEW

**Comment by zivnevo** [comment]
@elevran PTAL

**Comment by elevran** [comment]
@zivnevo thanks for the update. Pushed another iteration with more content. Might almost be there for initial sharing and feedback.

File: design-proposals/policy-attributes.md [hunk 1]
```diff
@@ -0,0 +1,281 @@
...
+
+### Prior discussion and links
+
+Not applicable.
+
+## User/User Story
+
+- **Access control based on cluster geography**: As a network administrator I would like to enable
+ Service access only from certain locations (e.g., EU only, to comply with GDPR).
+- **Load balancing based on cluster geography**: As a network administrator I would like to set
```
[the remark this answers is not in the document]
Comment (elevran) [reply design-proposals/policy-attributes.md:54]: My intent for user stories was to provide concrete examples of use cases motivating the needs. Thus, I tried to be specific in the examples, beyond "use attributes for policies".

File: design-proposals/policy-attributes.md [hunk 2]
```diff
@@ -0,0 +1,281 @@
...
+
+- Site level attributes (either a fixed set defined by ClusterLink or extended by user according to
+ the Fabric configuration [1]) that are pertinent to all workloads and Services in the site. Examples
+ may include `geography`, `cloud-provider`, `cloud-region`, `cluster-name`, etc.
+- Service level attributes (either a fixed set or augmented by user in the fabric configuration).
+ These may includes such attributes as `service-name`, `namespace`, `labels`, etc. Other attributes may
+ be derived from the Kubernetes Service definition, if relevant. Services are assigned the Site
+ attributes as well.
+- Workload attributes are associated with a specific workload instance, and may include, for example,
+ `service-account`, `namespace`, `image-name`, etc. Workloads are assigned the Site attributes as well.
```
[the remark this answers is not in the document]
Comment (elevran) [reply design-proposals/policy-attributes.md:110]: Please take a look at the latest version. I clarified (first paragraph under the `Proposal` section) the difference between Workload (client side) and Service (server side).

File: design-proposals/policy-attributes.md [hunk 3]
```diff
@@ -0,0 +1,304 @@
...
+ Service access only from certain locations (e.g., EU only, to comply with GDPR).
+- **Load balancing based on cluster geography**: As a network administrator I would like to set
+ a policy that, when a Service is provided by several remote locations, only locations with the
+ same geography as the source should be considered.
+- **Access control based on cluster identity**: As a Service owner, I would like to allow
+ access to a specific service only from other clusters I own.
+- **Access control based on workload namespace and labels**: As a Service owner, I would
+ like to enable access to a service based on the source workload namespace and its "role" label
+ value, regardless of cluster where the workload is running (e.g., assumes clusters are used as the
+ infrastructure and teams are allocated the same namespace across all clusters). As a user I would
```
Comment (michalmalka) [inline design-proposals/policy-attributes.md:65]: I would like...

**Comment by elevran** [comment]
@zivnevo - is this ready to be reviewed from your point of view?

**Comment by zivnevo** [comment]
> @zivnevo - is this ready to be reviewed from your point of view?

Yes, I think we are good

**Comment by gilshurek** [comment]
1.	In reference to identity-based-connectivity and access-control concepts: clarify which attribute-scope’ hold source/target attributes which should be tied to the identities of source/target entities (workload & service) – and which scopes hold ‘context’ attributes, preferably tied to identities of infrastructure (?) entities (gateway, site, fabric).
2.	Consistently use prefixes to denote attribute-scopes and sub-scopes.
a.	Why do we have a “cl” prefix? Is this the ‘gateway’ attribute-scope?
b.	Use different prefixes for Service and Workload k8s labels. 
c.	In general, I suggest not to use ‘k8s’ as a prefix but as a sub-prefix. E.g., “service:k8s: …” “workload:k8s: …” 
3.	The Image SHA as a candidate workload attribute: The SHA is a prime identity authentication attribute (as a spire selector) but maybe not an attribute to be used by the ‘connectivity-autorization’ policies: Once the workload’s identity is established, the policy engine will use connectivity decision attributes that are derived from the identity.
4.	“(note that the source site attributes are not sent to conserve resources - see note [here](https://github.com/clusterlink-net/clusterlink/blob/df3705ab0d5bc31f7526b26283784c8a4a1f5abf/design-proposals/policy-attributes.md#exchanging-attributes-between-gateways))”: I faild to find the note.
5.	Beyond this proposal: (how) Can we shift to a client handle which is not the IP-address?
6.	Beyond this proposal: We use a ‘service’ abstraction for the connection target, which may indicate a collection of workloads. How about using a similar abstraction for the source, including the requirement to ‘register’ such services to CL?

**Review by mrsabath — COMMENTED** [review]
This is a really good document that starts the conversation. I think we should have more discussion, as I started to add comments below.

File: design-proposals/policy-attributes.md [hunk 4]
```diff
@@ -0,0 +1,306 @@
...
+
+ClusterLink policies apply to communications between workloads. [ZN: should we mention services here?]
+ Workloads can be identified by a strong (e.g., cryptographic) identity.
+ The identity links the workload to a set of attributes, and policies are
+ defined on workload attributes. This design proposal defines the initial
+ set of attributes used in policies.
+
+ > ZN: The abstract only mentions workloads, but the notion of services is central to this document.
+ Should we mention services here?
+
```
[the remark this answers is not in the document]
Comment (zivnevo) [reply design-proposals/policy-attributes.md]: done

File: design-proposals/policy-attributes.md [hunk 5]
```diff
@@ -0,0 +1,306 @@
...
+- **Access control based on cluster identity**: As a Service owner, I would like to allow
+ access to a specific service only from other clusters I own.
+- **Access control based on workload namespace and labels**: as a Service owner, I would
+ like to enable access to a service based on the source workload namespace and its "role" label
+ value, regardless of cluster where the workload is running (e.g., assumes clusters are used as the
+ infrastructure and teams are allocated the same namespace across all clusters). Consequently, I would like
+ to enforce an egress policy that allows only workloads from *namespaces I own* on remote clusters and
+ not from namespaces assigned to other users. The labels of workloads running in other namespaces are
+ not trusted.
+
```
[the remark this answers is not in the document]
Comment (zivnevo) [reply design-proposals/policy-attributes.md:67]: added

File: design-proposals/policy-attributes.md [hunk 6]
```diff
@@ -0,0 +1,306 @@
...
+then the following attributes can be used to identify a workload (K8s Pod) within a Site:
+
+- K8s namespace
+- Other metadata fields, including
+  - Pod labels
+  - Pod name?
+  - Owner reference
+- Pod Spec fields, including
+  - Service Account
+  - Image name
```
[the remark this answers is not in the document]
Comment (zivnevo) [reply design-proposals/policy-attributes.md]: added

File: design-proposals/policy-attributes.md [hunk 7]
```diff
@@ -0,0 +1,306 @@
...
+
+> ZN: How should we set Image name if the Pod runs multiple containers?
+
```
[the remark this answers is not in the document]
Comment (zivnevo) [reply design-proposals/policy-attributes.md:146]: added

File: design-proposals/policy-attributes.md [hunk 8]
```diff
@@ -0,0 +1,306 @@
...
+### Exchanging Attributes Between Gateways
+
+For simplicity, let's assume that all gateways keep the attributes of all other gateways.
+Moreover, all gateways keep the attributes of all imported/exported services.
+While this assumption allows optimizing bandwidth and simplifies the below description, it is not
+strictly required. For this assumption to hold, management layer will probably need a mechanism
+to allow updating the attributes of gateways/services across the fabric.
+
+**Client Side:**
+
```
[the remark this answers is not in the document]
Comment (zivnevo) [reply design-proposals/policy-attributes.md:193]: added

**Comment by elevran** [comment]
@zivnevo can this be merged?

**Review by zivnevo — APPROVED** [review]

**Pull request merged**